src/routes/snippets.rs

```diff
@@ -77,6 +131,68 @@ pub fn create_snippet(
     create_snippet_impl(&**storage, &snippet, origin.path())
 }
 
+#[allow(clippy::too_many_arguments)]
+#[get("/snippets?<title>&<syntax>&<tag>&<marker>&<limit>")]
+pub fn list_snippets<'o, 'h>(
+    storage: State<Box<dyn Storage>>,
+    origin: &'o Origin,
+    title: Option<String>,
+    syntax: Option<String>,
+    tag: Option<String>,
+    limit: Option<PaginationLimit>,
+    marker: Option<String>,
+    _content_type: NegotiatedContentType,
+    _user: BearerAuth,
+) -> Result<WithHttpHeaders<'h, Output<Vec<Snippet>>>, ApiError> {
+    let mut criteria = ListSnippetsQuery {
+        title,
+        syntax,
+        tags: tag.map(|v| vec![v]),
+        ..Default::default()
+    };
+
+    // Fetch one more record in order to detect if there's a next page, and
+    // generate appropriate Link entry accordingly.
+    criteria.pagination.limit = limit.map(|v| v.0).unwrap_or(criteria.pagination.limit) + 1;
+    criteria.pagination.marker = marker;
+
+    let limit = criteria.pagination.limit - 1;
+    let mut snippets = storage.list(criteria.clone())?;
+    let mut prev_needed = false;
+    let next_marker = if snippets.len() > limit {
```

Comment:
I'll see what I can do with moving these lines into helper function. 

> snippets.pop()

We can't use it because we're interested in the item before last, not the last one.

Reply:
Argh, I missed the fact that we are doing -1 twice :(